src/Functions/GatherUtils/Selectors.h

```diff
@@ -115,8 +115,11 @@ struct ArraySourcePairSelector
     static void selectSource(bool is_second_const, bool is_second_nullable, SecondSource && second,
                              bool is_first_const, bool is_first_nullable, FirstSource && first, Args && ... args)
     {
-        Base::selectSourcePair(is_first_const, is_first_nullable, first,
-                               is_second_const, is_second_nullable, second, args ...);
+        if constexpr (std::is_same_v<FirstSource, SecondSource>)
```

Comment:
The main change is here: we were calling `selectSource` only for columns of identical types, but the code monomorphisation was done for every possible pair of column types.